### ISSUE TITLE
Async fan-out Scenarios?

### ISSUE THREAD

**Issue by ideepg** [grpc/grpc-go#3280]
Any examples of Async fan-out scenarios ? Is is possible today? If yes how?

**Comment by menghanl** [comment]
Can you give more information on what you are asking about? An example of the scenario you are asking about would be helpful.

**Comment by gullyy** [comment]
Here is the scenario.Below mentioned are both grpc server and client

Client Calls to the Server. Client is interested in a topic. 
Server sends a response to the Client that your request has been accepted. As soon as there is data on the topic you(Client) will be notified.
Every time server receives a data on the topic(how does the server receive the data on the topic is out of scope.), it is sent across the client.

Question 
1. I understand grpc is pretty sync . To me the above flow is async in nature. Is the above flow possible with grpc?
2. I hear about [Completion Queue](https://grpc.github.io/grpc/core/md_doc_core_grpc-cq.html) . Is completion queue a mechanism by which this problem can be solved? I fail to find completion queue implementation/example  in Golang .

To summarise - grpc works for req-res paradigm. Does it also work for req-res-res-res...res(to multiple client- aka fan out) paradigm ?If yes  where are the examples?

**Comment by gullyy** [comment]
One quick way to achieve the above scenario would be the following.

1. The client calls the server, the client is interested in a topic.
2. The server acknowledges the request. Stores the topic name and grpc-server co-ordinates which the client has passed
3. The client is expected to run its own grpc-server. This is a part of the contract
4. The server now receives data at the topic. Server looks up who is/are  interested at this topic.
5. Figures out there are 2 grpc-server who are interested in this topic.
6. Server initiates a grpc connection to these grpc-servers- in turn it becomes a client. Sends the data to these servers.

Problem with this approach is - too much of glue code and extra management of co-ordinates and connections. Looking for a better approach to solve this problem.

**Comment by menghanl** [comment]
It seems what you want is https://grpc.io/docs/guides/concepts/#server-streaming-rpc

An example: [proto definition](https://github.com/grpc/grpc-go/blob/aa428f8ea4a8cdc385472d23116438df3344d07e/examples/route_guide/routeguide/route_guide.proto#L39), [client](https://github.com/grpc/grpc-go/blob/aa428f8ea4a8cdc385472d23116438df3344d07e/examples/route_guide/client/client.go#L59), [server](https://github.com/grpc/grpc-go/blob/aa428f8ea4a8cdc385472d23116438df3344d07e/examples/route_guide/server/server.go#L78).
